### PR TITLE
How to clear cache after pruning images

### PR DESCRIPTION
**Pull request by mburke5678**
https://bugzilla.redhat.com/show_bug.cgi?id=1730148

Per Slack: https://coreos.slack.com/archives/CEKNRGF25/p1573081260097100?thread_ts=1573079169.094900&cid=CEKNRGF25

### CODE REVIEW

**Comment by mburke5678** [comment]
@adambkaplan @wking PTAL

File: modules/pruning-images.adoc [hunk 1]
```diff
@@ -28,20 +28,39 @@ By default, the integrated registry caches blobs metadata to reduce the number
 of requests to storage, and increase the speed of processing the request.
 Pruning does not update the integrated registry cache. Images pushed after
 pruning that contain pruned layers will be broken, because the pruned layers
-that have metadata in the cache will not be pushed. Therefore, it is necessary
-to clear the cache after pruning. This can be accomplished by redeploying the
-registry:
+that have metadata in the cache will not be pushed. To avoid this issue,
+you can set the image registry to `readOnly: true` before pruning and 
+reset to `false` when complete.
```
Comment (mburke5678) [inline modules/pruning-images.adoc]: @adambkaplan When would the use reset to `false`?

File: modules/pruning-images.adoc [hunk 2]
```diff
@@ -28,20 +28,39 @@ By default, the integrated registry caches blobs metadata to reduce the number
...
 
 ----
-$ oc rollout latest dc/docker-registry
+$ oc edit configs.imageregistry.operator.openshift.io/cluster
```
Comment (wking) [inline modules/pruning-images.adoc]: nit: `oc patch ...` so you don't have to list the unrelated stuff like proxy settings?

File: modules/pruning-images.adoc [hunk 3]
```diff
@@ -28,20 +28,39 @@ By default, the integrated registry caches blobs metadata to reduce the number
...
 pruning that contain pruned layers will be broken, because the pruned layers
-that have metadata in the cache will not be pushed. Therefore, it is necessary
-to clear the cache after pruning. This can be accomplished by redeploying the
-registry:
+that have metadata in the cache will not be pushed. To avoid this issue,
+you can set the image registry to `readOnly: true` before pruning and 
+reset to `false` when complete.
 
 ----
-$ oc rollout latest dc/docker-registry
```
Comment (wking) [inline modules/pruning-images.adoc]: nit: do we want to remove this for master?  I thought this would work, and still be preferred, for 4.3+.  But I'm unfamiliar with the registry internals.

Comment (adambkaplan) [reply modules/pruning-images.adoc]: @wking this was from 3.x. OpenShift 4 has completely different namings for the deployment

Comment (mburke5678) [reply modules/pruning-images.adoc]: @wking I was planning to merge this PR to master and cherrypick to 4.1 and 4.2 for now. Then merge https://github.com/openshift/openshift-docs/pull/17966/files to master when 4.3 is ready to go. The 4.3 PR has: `$ oc rollout restart deployment/image-registry -n openshift-image-registry`

**Comment by adambkaplan** [comment]
@mburke5678 @wking I did the following to confirm that the following will work:

For 4.3 - `oc rollout restart deployment image-registry -n openshift-image-registry`

For 4.2 and 4.1 - `oc patch deployment image-registry -n openshift-image-registry --type=merge --patch='{"spec":{"template":{"metadata":{"annotations":{"kubectl.kubernetes.io/restartedAt": "restartTimestamp"}}}}}'` - note that the annotation value can be anything, we just prefer a UTC timestamp (it just needs to change with each restart)

**Comment by adambkaplan** [comment]
That said, the reason this works cleanly is because of a bug: https://bugzilla.redhat.com/show_bug.cgi?id=1769927

These commands can force up to _two_ rollouts on other operator-controlled deployments - one that adds the restart annotation, and a second rollout that the operator initiates to remove the annotation.

/cc @soltysh @sttts 

Is the race that may trigger two rollouts something we should worry about (ex - for ocm or the apiservers)?

**Comment by mburke5678** [comment]
@adambkaplan Is the `oc rollout restart`subcommand in the `oc` yet?

```
$ oc version
Client Version: openshift-clients-4.3.0-201910250623-42-gc276ecb7
Server Version: 4.3.0-0.nightly-2019-11-08-094604
Kubernetes Version: v1.16.2
~ $ oc rollout restart
error: unknown command "restart"
See 'oc rollout -h' for help and examples
```

**Comment by adambkaplan** [comment]
@tnozicka is this the correct command upstream? Do we need to file a BZ?

**Comment by mburke5678** [comment]
@tnozicka is this the correct command upstream? Do we need to file a BZ?

**Comment by tnozicka** [comment]
yes, file a BZ to wire it

**Comment by tnozicka** [comment]
assign the BZ to me, fyi the fix is here https://github.com/openshift/oc/pull/169

File: modules/pruning-images.adoc (hunk 1 above)
Comment (mburke5678) [reply modules/pruning-images.adoc]: @adambkaplan ^^
